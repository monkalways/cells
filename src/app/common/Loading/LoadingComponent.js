import React from 'react';
import PropTypes from 'prop-types';
import { CircularProgress, withStyles } from '@material-ui/core';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  size: PropTypes.number,
};

const defaultProps = {
  size: 100,
};

const LoadingComponent = ({ classes, size }) => (
  <div className={classes.container}>
    <CircularProgress className={classes.progress} size={size} />
  </div>
);

LoadingComponent.propTypes = propTypes;
LoadingComponent.defaultProps = defaultProps;

export default withStyles((theme) => ({
  container: {
    display: 'flex',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  progress: {
    margin: theme.spacing.unit * 2,
  },
}))(LoadingComponent);
