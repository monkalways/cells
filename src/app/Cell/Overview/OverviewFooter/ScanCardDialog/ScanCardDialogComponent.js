import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
  withStyles,
} from '@material-ui/core';

import commonConstants from '../../../../constants';
import Loading from '../../../../common/Loading';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  isSignInDialogOpen: PropTypes.bool.isRequired,
  isAuthenticating: PropTypes.bool.isRequired,
  isAuthenticationFailed: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  authenticate: PropTypes.func.isRequired,
};

class ScanCardDialogComponent extends Component {
  state = {
    cardId: '',
  };

  componentWillMount() {
    document.addEventListener('keydown', this.handleKeyDown.bind(this));
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyDown.bind(this));
  }

  handleKeyDown = (event) => {
    const { authenticate } = this.props;
    let { cardId } = this.state;
    cardId += event.key;
    this.setState({
      cardId,
    });

    if (cardId.length === commonConstants.SCAN_CARD_ID_LENGTH) {
      authenticate(cardId);
      this.setState({
        cardId: '',
      });
    }
  };

  handleFakeUserClick = () => {
    const { authenticate } = this.props;
    authenticate('b5a3d600f8ff12e0');
  };

  render() {
    const {
      classes,
      isSignInDialogOpen,
      isAuthenticating,
      isAuthenticationFailed,
      onClose,
    } = this.props;
    return (
      <Dialog open={isSignInDialogOpen} onClose={onClose}>
        <DialogTitle id="alert-dialog-title">Sign In</DialogTitle>
        <DialogContent className={classes.dialog}>
          {isAuthenticating ? (
            <Loading size={50} />
          ) : isAuthenticationFailed ? (
            <DialogContentText id="alert-dialog-description" variant="body1">
              Authentication Failed
            </DialogContentText>
          ) : (
            <DialogContentText id="alert-dialog-description" variant="body1">
              Please scan your ID card to proceed ...
            </DialogContentText>
          )}
        </DialogContent>
        <DialogActions>
          {
            // TODO: Remove the "Proceed with Test Account" button
          }
          <Button
            onClick={this.handleFakeUserClick}
            color="secondary"
            autoFocus
          >
            Proceed with Test Account
          </Button>
          <Button onClick={onClose} color="primary" autoFocus>
            Cancel
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

ScanCardDialogComponent.propTypes = propTypes;

export default withStyles((theme) => ({
  dialog: {
    width: theme.spacing.unit * 50,
  },
}))(ScanCardDialogComponent);
