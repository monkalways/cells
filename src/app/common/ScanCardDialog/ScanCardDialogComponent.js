import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
  withStyles,
} from '@material-ui/core';

import commonConstants from '../../constants';
import Loading from '../Loading';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  isSignInDialogOpen: PropTypes.bool.isRequired,
  isAuthenticating: PropTypes.bool.isRequired,
  isAuthenticationFailed: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  authenticate: PropTypes.func.isRequired,
  handleSignIn: PropTypes.func.isRequired,
};

export class ScanCardDialogComponent extends Component {
  state = {
    cardId: '',
  };

  componentDidMount() {
    window.addEventListener('keydown', this.handleKeyDown.bind(this));
  }

  componentWillUnmount() {
    this.isCancelled = true;
    window.removeEventListener('keydown', this.handleKeyDown.bind(this));
  }

  handleKeyDown = (event) => {
    if (this.isCancelled) return;

    const { isSignInDialogOpen, handleSignIn, authenticate } = this.props;
    if (!isSignInDialogOpen) handleSignIn();

    let { cardId } = this.state;
    cardId += event.key;
    this.setState({
      cardId,
    });

    if (cardId.length === commonConstants.SCAN_CARD_ID_LENGTH) {
      authenticate(cardId);
      this.setState({
        cardId: '',
      });
    }
  };

  handleFakeUserClick = () => {
    const { authenticate } = this.props;
    authenticate('b5a3d600f8ff12e0');
  };

  render() {
    const {
      classes,
      isSignInDialogOpen,
      isAuthenticating,
      isAuthenticationFailed,
      onClose,
    } = this.props;
    return (
      <Dialog open={isSignInDialogOpen} onClose={onClose}>
        <DialogTitle id="alert-dialog-title">Sign In</DialogTitle>
        <DialogContent className={classes.dialog}>
          {isAuthenticating ? (
            <Loading size={50} />
          ) : isAuthenticationFailed ? (
            <DialogContentText id="alert-dialog-description" variant="body1">
              Authentication Failed
            </DialogContentText>
          ) : (
            <DialogContentText id="alert-dialog-description" variant="body1">
              Please scan your ID card to proceed ...
            </DialogContentText>
          )}
        </DialogContent>
        <DialogActions>
          {
            // TODO: Remove the "Proceed with Test Account" button
          }
          <Button
            id="fakeUserLoginButton"
            onClick={this.handleFakeUserClick}
            color="secondary"
            autoFocus
          >
            Proceed with Test Account
          </Button>
          <Button onClick={onClose} color="primary" autoFocus>
            Cancel
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

ScanCardDialogComponent.propTypes = propTypes;

export default withStyles((theme) => ({
  dialog: {
    width: theme.spacing.unit * 50,
  },
}))(ScanCardDialogComponent);
