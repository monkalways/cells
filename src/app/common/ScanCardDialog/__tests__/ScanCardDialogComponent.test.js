import React from 'react';
import { shallow } from 'enzyme';

import { DialogContentText, Dialog } from '@material-ui/core';
import { ScanCardDialogComponent } from '../ScanCardDialogComponent';
import Loading from '../../Loading';
import commonConstants from '../../../constants';

describe('ScanCardDialogComponent', () => {
  let classes;
  let isSignInDialogOpen;
  let isAuthenticating;
  let isAuthenticationFailed;
  let onClose;
  let authenticate;
  let handleSignIn;

  beforeEach(() => {
    classes = {
      dialog: 'dialog',
    };
    isSignInDialogOpen = false;
    isAuthenticating = false;
    isAuthenticationFailed = false;
    onClose = jest.fn();
    authenticate = jest.fn();
    handleSignIn = jest.fn();
  });

  const setup = () => shallow(<ScanCardDialogComponent
    classes={classes}
    isSignInDialogOpen={isSignInDialogOpen}
    isAuthenticating={isAuthenticating}
    isAuthenticationFailed={isAuthenticationFailed}
    onClose={onClose}
    authenticate={authenticate}
    handleSignIn={handleSignIn}
  />);

  it('should render "Please scan your ID card to proceed" if not authenticating', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(DialogContentText).html()).toContain('Please scan your ID card to proceed');
  });

  it('should render "Authentication Failed" if not authenticating but authentication failed', () => {
    isAuthenticationFailed = true;
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(DialogContentText).html()).toContain('Authentication Failed');
  });

  it('should render loading component if authenticating', () => {
    isAuthenticating = true;
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(Loading)).toExist();
  });

  it('should add keydown event listener when componentDidMount is called, and remove event listener when componentWillUnmount is called', () => {
    const map = {};
    window.addEventListener = jest.fn((event, eventListener) => {
      map[event] = eventListener;
    });
    window.removeEventListener = jest.fn((event) => {
      map[event] = null;
    });
    const wrapper = setup();

    wrapper.instance().componentDidMount();
    map.keydown({ key: 'A' });

    expect(handleSignIn).toBeCalled();

    wrapper.instance().componentWillUnmount();
    expect(map.keydown).toBe(null);
    expect(wrapper.instance().isCancelled).toBe(true);
  });

  describe('handleKeyDown', () => {
    let map;

    beforeEach(() => {
      map = {};
      window.addEventListener = jest.fn((event, eventListener) => {
        map[event] = eventListener;
      });
      window.removeEventListener = jest.fn((event) => {
        map[event] = null;
      });
    });

    it('should do nothing if isCancelled is true', () => {
      const wrapper = setup();
      wrapper.instance().isCancelled = true;

      map.keydown({ key: 'A' });

      expect(wrapper.state('cardId')).toBe('');
    });

    it('should call handleSignIn if isSignInDialogOpen is false', () => {
      setup();

      map.keydown({ key: 'A' });

      expect(handleSignIn).toBeCalled();
    });

    it('should not call handleSignIn if isSignInDialogOpen is true', () => {
      isSignInDialogOpen = true;
      setup();

      map.keydown({ key: 'A' });

      expect(handleSignIn).not.toBeCalled();
    });

    it('should collect card id from keydown events', () => {
      const wrapper = setup();

      map.keydown({ key: 'A' });
      map.keydown({ key: 'B' });
      map.keydown({ key: 'C' });
      map.keydown({ key: '1' });
      map.keydown({ key: '2' });
      map.keydown({ key: '3' });

      expect(wrapper.state('cardId')).toBe('ABC123');
    });

    it(`should reset card id and call authencate if ${
      commonConstants.SCAN_CARD_ID_LENGTH
    } characters are collected from keydown events`, () => {
      const wrapper = setup();

      map.keydown({ key: 'A' });
      map.keydown({ key: 'B' });
      map.keydown({ key: 'C' });
      map.keydown({ key: 'D' });
      map.keydown({ key: 'E' });
      map.keydown({ key: 'F' });
      map.keydown({ key: 'G' });
      map.keydown({ key: '1' });
      map.keydown({ key: '2' });
      map.keydown({ key: '3' });
      map.keydown({ key: '4' });
      map.keydown({ key: '5' });
      map.keydown({ key: '6' });
      map.keydown({ key: '7' });
      map.keydown({ key: '8' });
      map.keydown({ key: '9' });

      expect(authenticate).toBeCalledWith('ABCDEFG123456789');
      expect(wrapper.state('cardId')).toBe('');
    });
  });

  it('should handle fake users click for now (TO BE REMOVED)', () => {
    const wrapper = setup();
    wrapper.find('#fakeUserLoginButton').simulate('click');

    expect(authenticate).toBeCalledWith('b5a3d600f8ff12e0');
  });

  it('should handle dialog close', () => {
    const wrapper = setup();
    wrapper.find(Dialog).simulate('close');

    expect(onClose).toBeCalled();
  });
});
