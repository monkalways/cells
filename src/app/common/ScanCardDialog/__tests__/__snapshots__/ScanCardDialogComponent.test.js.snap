// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanCardDialogComponent should render "Authentication Failed" if not authenticating but authentication failed 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={[MockFunction]}
  open={false}
>
  <WithStyles(DialogTitle)
    id="alert-dialog-title"
  >
    Sign In
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)
    className="dialog"
  >
    <WithStyles(DialogContentText)
      id="alert-dialog-description"
      variant="body1"
    >
      Authentication Failed
    </WithStyles(DialogContentText)>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[MockFunction]}
    >
      Cancel
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`ScanCardDialogComponent should render "Please scan your ID card to proceed" if not authenticating 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={[MockFunction]}
  open={false}
>
  <WithStyles(DialogTitle)
    id="alert-dialog-title"
  >
    Sign In
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)
    className="dialog"
  >
    <WithStyles(DialogContentText)
      id="alert-dialog-description"
      variant="body1"
    >
      Please scan your ID card to proceed ...
    </WithStyles(DialogContentText)>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[MockFunction]}
    >
      Cancel
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`ScanCardDialogComponent should render loading component if authenticating 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={[MockFunction]}
  open={false}
>
  <WithStyles(DialogTitle)
    id="alert-dialog-title"
  >
    Sign In
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)
    className="dialog"
  >
    <WithStyles(LoadingComponent)
      size={50}
    />
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[MockFunction]}
    >
      Cancel
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
